### DISCUSSION TITLE
Dual vs. Single Stack

### DISCUSSION THREAD

**Discussion by jenslink** [becarpenter/book6#67] [General]
I'll start working on a new old IPv6 project shortly, and their current plan is to enable dual-stack everywhere and then go from there to IPv6 only in the coming year. From my point of view forcing dual-stack makes more work, think configuration, monitoring, troubleshooting, and should be avoided if possible. So kind of dual-stack where you must, single stack IPv6 where you can and maybe keep IPv4 only for things that have to be replaced in the (near) future anyway. 

I think there should be something about this in the book.

**Comment by becarpenter** [comment]
> plan is to enable dual-stack everywhere and then go from there to IPv6 only in the coming year

That seems incredibly optimistic. I've always been a fan of traditional dual stack when there is a large _operational_ IPv4 legacy that will likely take years to clear. On the other hand, going straight to IPv4-as-a-service is a bigger initial workload but at least in theory significantly reduces OPEX.

We do cover this at [1] and [2] but what we really need is a case study in Chapter 7. There is always some confusion, because until IPv4 is truly dead (i.e. for ever, as far as I can see), apps still need to see a dual-stack socket API, but with IPv4-as-a-service there are no native IPv4 packets on the wire so no dual-stack routing. Anyway, how about writing a case study based on your new project?

[1] https://github.com/becarpenter/book6/blob/main/3.%20Coexistence%20with%20Legacy%20IPv4/3.%20Coexistence%20with%20Legacy%20IPv4.md
[2] https://github.com/becarpenter/book6/blob/main/3.%20Coexistence%20with%20Legacy%20IPv4/Dual%20stack%20scenarios.md

**Reply by jenslink** [reply]
I know of one large enterprise which is really short on IPv4 addresses, including RFC1918. For them, it was a win to move 1000+ printers to IPv6 only and so get free IPv4 addresses.

**Reply by buraglio** [reply]
We've taken that approach in several places, Jen. Brian is correct, it's pretty optimistic, but it does short-cut and save a LOT of time when it does work. The hill is more steep, but the ride down from the peak is significantly less rocky.

**Reply by defdefred** [reply]
How a company could be short on ipv4 internally?
10.0.0.0 to 10.255.255.255 is free for everybody: 16M IPs...

**Reply by ehorley** [reply]
My company works with Fortune 500 and Large US Federal Departments and Agencies. I would say the majority of those we work with are very short or have completely run out of IPv4 internally for 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, 100.64.0.0/10, and are even squatting on public IPv4 space from DOD and others that they hope will not be an issue anytime soon. It is a real-world problem, and we have seen IPv6 deployments to address this issue alone. It may not be an issue everyone is seeing, but some very significant and important companies are having to deal with it, so it would be good to have a roadmap and tactics that cover what needs to be done to address these needs. Just me 2 cents.

**Reply by becarpenter** [reply]
Not to mention that when two companies or agencies using Net 10 merge, they have an immediate crisis due to address collisions. But @defdefred is correct when it comes to smaller enterprises, and they may stay on IPv4 indefinitely, I'm afraid, unitl the incentives change for other reasons.

**Reply by ehorley** [reply]
Yes, smaller enterprises have fewer pressure points to drive their behavior. Perhaps the egos of these technical executive teams to look/act/operate like their bigger competitors, but I wouldn't pin any hopes of adoption on that angle. It will be a long slow slog up to 75% adoption. At that point, there might be a factor of "we are being left behind," which could push the laggards to adopt. By that time, the services and the rest of the Internet will likely all be IPv6 first, so that might be an additional consideration too. I don't see a lot of folks bragging about still being on AppleTalk or IPX/SPX these days but I am sure they are out there...

**Reply by jenslink** [reply]
> even squatting on public IPv4 space from DOD and others that they hope will not be an issue anytime soon.

44/8 is a good example. I know one company, and heard of at least two others, that are using this. IIRC one of those is also an AWS customer. (A large chunk of 44/8 is assigned to AWS).

**Reply by jenslink** [reply]
> How a company could be short on ipv4 internally? 10.0.0.0 to 10.255.255.255 is free for everybody: 16M IPs...

It's easy to use all of RFC1918 space.  Networks are not a new thing. They've grown and evolved over the years, people were very generous assigning prefixes. Add mergers and acquisitions to the mix, and you have a high probability of using the same address space more than once inside your network. Often documentation is lacking. Cleaning up your IPv4 network is a lot of work and has a high potential of breaking things.

**Reply by defdefred** [reply]
Ok. Messy companies could waste 16M IP addresses...
Maybe they could learn to use an IPAM software and take time to update their documentation...
Company like this deserve to have issue.
Cleaning the IT is the full time job of IT guys...
If they struggle to manage the simple IPv4, how could they manage IPv6?

**Comment by becarpenter** [comment]
Perfect candidate for a case study for chapter 7! Get writing??
